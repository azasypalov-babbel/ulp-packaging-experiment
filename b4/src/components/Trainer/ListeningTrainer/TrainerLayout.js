import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import StyledTrainerLayout from '../../TrainerLayout';

const Section = styled.section`
  display: flex;
  flex-direction: column;
  align-items: center;

  &:nth-child(1) {
    padding: 2rem;
    flex: 0 0 auto;
  }

  &:nth-child(2) {
    padding: 1rem;
    flex: 1 0 auto;
    justify-content: center;
  }
`;

const TrainerLayout = ({ top, bottom }) => (
  <StyledTrainerLayout>
    <Section>{top}</Section>
    <Section>{bottom}</Section>
  </StyledTrainerLayout>
);

TrainerLayout.propTypes = {
  top: PropTypes.node,
  bottom: PropTypes.node
};

export default TrainerLayout;
