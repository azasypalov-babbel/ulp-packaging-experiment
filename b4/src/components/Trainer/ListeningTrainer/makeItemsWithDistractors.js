function sortByTextLength(items) {
  return items.sort(function(a, b) {
    const lengthA = a.learnLanguageText.length;
    const lengthB = b.learnLanguageText.length;
    if (lengthA < lengthB) {
      return -1;
    }
    if (lengthA > lengthB) {
      return 1;
    }
    return 0;
  });
}

function chunkItems(items, chunkSize) {
  let results = [];

  let i = 0;
  while (i < items.length) {
    results.push(items.slice(i, i + chunkSize));
    i += chunkSize;
  }

  return results;
}

export function applyGroupingStrategy(items) {
  if (items.length < 6) {
    return [items];
  }

  const sortedItems = sortByTextLength(items);

  let result;
  if (sortedItems.length === 6) {
    result = chunkItems(sortedItems, 3);
  } else if (sortedItems.length < 9) {
    result = chunkItems(sortedItems, 4);
  } else if (sortedItems.length % 5 == 0 || sortedItems.length % 5 > 2) {
    result = chunkItems(sortedItems, 5);
  } else {
    result = chunkItems(sortedItems, 4);
  }


  return result;
}

function removeCurrentItem(items, itemIndex) {
  return items.filter(function(item, index) {
    return index !== itemIndex;
  });
}

function buildDistractors(items) {
  return items.map(function(item, i) {
    return {
      item: item,
      distractors: removeCurrentItem(items, i)
    };
  });
}

function flattenItemsReducer(acc, val) {
  return acc.concat(val);
}

export default function makeItemsWithDistractors(items) {
  const itemsWithDistractors = applyGroupingStrategy(items)
    .map(buildDistractors)
    .reduce(flattenItemsReducer, []);

  return itemsWithDistractors;
}
