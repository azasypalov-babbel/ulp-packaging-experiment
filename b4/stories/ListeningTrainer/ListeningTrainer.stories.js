import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { boolean } from '@storybook/addon-knobs';
import { ListeningTrainer } from '../../src/components/Trainer/ListeningTrainer/ListeningTrainer';

import { withMockServicesProvider } from '../decorators/withServices';
import withStore from '../decorators/withStore';
import withAudio from '../../src/components/shared/withAudio';

const stories = storiesOf('ListeningTrainer', module)
  .addDecorator(withStore)
  .addDecorator(withMockServicesProvider);

const ListeningTrainerWithAudio = withAudio()(ListeningTrainer);

const getProps = () => ({
  item: {
    id: '549f9ecf012e7f3b636ec9e4ac5a0d10',
    learnLanguageText: 'Eins',
    displayLanguageText: 'One',
    sound: {
      id: 'd3572193477785b1bc030cdf61ec777c'
    }
  },
  showChoiceItems: boolean('showChoiceItems', false),
  onUserTriggeredSound: action('onUserTriggeredSound'),
  onContinue: action('onContinue'),
  onChoiceItemClick: () => {},
  choiceItemsList: [
    { learnLanguageText: 'Zwei', displayLanguageText: 'Two' },
    { learnLanguageText: 'Drei', displayLanguageText: 'Three' },
    { learnLanguageText: 'Fear', displayLanguageText: 'Four' },
    { learnLanguageText: 'Funf', displayLanguageText: 'Five' }
  ],
  translations: {
    attemptFeedback: {
      negativeFeedback: 'Keep working on!',
      positiveFeedback: 'Fantastic!'
    },
    continueBtn: 'Continue',
    instruction: 'Listen to the phrase'
  },
  mediaUrlService: {
    soundURL: (val) => `http://localhost/${val}.mp3`
  },
  onSoundPlayerEnded: () => action('Sound player ended'),
  onSoundPlayerError: () => action('Sound player error'),
  onToggleShortcuts: action('onToggleShortcuts')
});

stories.add('ListeningTrainer', () => {
  const props = getProps();
  return <ListeningTrainerWithAudio {...props} />;
});

