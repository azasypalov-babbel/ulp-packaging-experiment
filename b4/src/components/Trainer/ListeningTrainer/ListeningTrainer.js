import React from 'react';
import PropTypes from 'prop-types';

import ChoiceList from './ChoiceList/ChoiceList';
import SoundPlayer from '../../shared/SoundPlayer';
import FeedbackSheet from '../../shared/FeedbackSheet/FeedbackSheet';
import { audioPropTypes } from '../../shared/withAudio';
import Title from '../../shared/Title';
import Toolbar from '../../shared/Toolbar';
import TrainerLayout from './TrainerLayout';
import IconButton from '../../shared/IconButton/IconButton';
import HighlightedText from '../../shared/FeedbackSheet/HighlightedText';

const isCorrectAttempt = (attempt) => attempt.mistakes === 0;

const getFeedbackMessage = (attemptFeedback, correct) =>
  correct ? attemptFeedback.positiveFeedback : attemptFeedback.negativeFeedback;

export class ListeningTrainer extends React.Component {
  constructor(props) {
    super(props);

    this.handleChoiceItemClick = this.handleChoiceItemClick.bind(this);
    this.handleContinueClick = this.handleContinueClick.bind(this);
  }

  handleChoiceItemClick(attemptedText, evt = {}) {
    const keyboardTriggered = evt.type !== 'click';
    const { attempt, item, onChoiceItemClick } = this.props;

    if (!attempt) {
      const itemText = item.displayLanguageText;
      const mistakes = (attemptedText === itemText) ? 0 : 2;

      onChoiceItemClick({
        mistakes,
        text: attemptedText
      }, keyboardTriggered);
    }
  }

  handleContinueClick() {
    const { attempt, onContinue, item } = this.props;

    if (!attempt) {
      // prevent user from clicking continue button several times
      return;
    }

    onContinue(item, attempt.mistakes);
  }

  render() {
    const {
      onSoundPlayerEnded,
      onSoundPlayerError,
      attempt,
      showChoiceItems,
      item,
      choiceItemsList,
      audio,
      onToggleShortcuts,
      translations
    } = this.props;

    const correctText = item.displayLanguageText;
    const choiceItemsTextList = choiceItemsList.map((item) => item.displayLanguageText);

    const highlightedText = (
      <div data-selector="learn-language-text">
        <HighlightedText appearance="POSITIVE">
          {item.learnLanguageText}
        </HighlightedText>
      </div>
    );

    return (
      <div>
        <TrainerLayout
          top={
            <>
              <Title data-selector="listening-title" fontSize="big">{translations.instruction}</Title>
              <SoundPlayer
                key={this.props.item.id}
                onEnded={onSoundPlayerEnded}
                onError={onSoundPlayerError}
                playOnMount
                audio={audio}
                url={this.props.mediaUrlService.soundURL(item.sound.id)}
              />
            </>
          }
          bottom={
            <ChoiceList
              showChoiceItems={showChoiceItems}
              items={choiceItemsTextList}
              attemptedText={attempt && attempt.text}
              correctText={correctText}
              onClick={this.handleChoiceItemClick}
            />
          }
        />
        {!attempt &&
          <Toolbar>
            <IconButton
              data-selector="reveal-shortcuts-button"
              iconName="KeyboardIcon"
              onClick={onToggleShortcuts}
            />
          </Toolbar>
        }
        {attempt &&
          <FeedbackSheet
            data-gap-feedback={true}
            data-solved={isCorrectAttempt(attempt)}
            data-mistaken={!isCorrectAttempt(attempt)}
            key="feedback-sheet"
            feedbackMessage={getFeedbackMessage(translations.attemptFeedback, isCorrectAttempt(attempt))}
            feedbackDetail={highlightedText}
            onContinueClick={this.handleContinueClick}
          />
        }
      </div>
    );
  }
}

ListeningTrainer.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.string.isRequired,
    sound: PropTypes.shape({
      id: PropTypes.string.isRequired
    }).isRequired,
    learnLanguageText: PropTypes.string.isRequired,
    displayLanguageText: PropTypes.string.isRequired
  }),
  choiceItemsList: PropTypes.array.isRequired,
  onContinue: PropTypes.func.isRequired,
  onChoiceItemClick: PropTypes.func.isRequired,
  audio: audioPropTypes.isRequired,
  attempt: PropTypes.shape({
    mistakes: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired
  }),
  mediaUrlService: PropTypes.shape({
    soundURL: PropTypes.func.isRequired
  }).isRequired,
  showChoiceItems: PropTypes.bool.isRequired,
  onSoundPlayerEnded: PropTypes.func.isRequired,
  onSoundPlayerError: PropTypes.func.isRequired,
  onToggleShortcuts: PropTypes.func.isRequired,
  translations: PropTypes.shape({
    attemptFeedback: PropTypes.shape({
      negativeFeedback: PropTypes.string.isRequired,
      positiveFeedback: PropTypes.string.isRequired
    }).isRequired,
    continueBtn: PropTypes.string.isRequired,
    instruction: PropTypes.string.isRequired
  }).isRequired
};

import pickRandom from '../../../lib/pickRandom';
import withTranslations from '../../shared/withTranslations';

const getTranslations = (translate, props, instanceId) => {
  const positiveFeedbackMessages = [
    'listening_trainer.feedback.positive.1',
    'listening_trainer.feedback.positive.2',
    'listening_trainer.feedback.positive.3',
    'listening_trainer.feedback.positive.4',
    'listening_trainer.feedback.positive.5',
    'listening_trainer.feedback.positive.6'
  ].map(translate);
  const negativeFeedbackMessages = [
    'listening_trainer.feedback.negative.1',
    'listening_trainer.feedback.negative.2',
    'listening_trainer.feedback.negative.3',
    'listening_trainer.feedback.negative.4',
    'listening_trainer.feedback.negative.5'
  ].map(translate);

  return {
    attemptFeedback: {
      positiveFeedback: pickRandom(positiveFeedbackMessages, instanceId),
      negativeFeedback: pickRandom(negativeFeedbackMessages, instanceId)
    },
    continueBtn: translate('listening_trainer.button.continue'),
    instruction: translate('listening_trainer.instruction.listen')
  };
};

export default withTranslations(getTranslations)(ListeningTrainer);
