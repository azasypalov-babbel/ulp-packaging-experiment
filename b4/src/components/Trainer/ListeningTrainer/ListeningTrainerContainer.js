import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import ListeningTrainer from './ListeningTrainer';

import * as sequenceActions from '../../../dux/sequence/actions';
import { toggleHints } from '../../../dux/keyboard/actions';
import { track } from '../../../dux/tracker/actions';
import { MESSAGE_KEYS } from '../../../dux/messages/messageKeys';
import * as messagesActions from '../../../dux/messages/actions';

import { withServices } from '../../shared/withServices';
import withAudio, { audioPropTypes } from '../../shared/withAudio';

import makeItemsWithDistractors from './makeItemsWithDistractors';
import { formatDL, formatLL } from '../../../lib/markupFormatter';
import shuffle from '../../../lib/shuffle';
import playFeedbackSound, { FEEDBACK_SOUND } from '../../../lib/playFeedbackSound';


const formatTrainerItem = (item) => ({
  id: item.id,
  sound: item.sound,
  learnLanguageText: formatLL(item.learn_language_text),
  displayLanguageText: formatDL(item.display_language_text)
});

const flattenItems = (trainer) => {
  return trainer.item_groups.reduce((acc, group) => {
    return acc.concat(group.items);
  }, []);
};

const shuffleItems = (currentIndex, items) => {
  const { item, distractors } = items[currentIndex];
  return shuffle({ array: [item, ...distractors] });
};

export class ListeningTrainerContainer extends React.Component {
  constructor(props) {
    super(props);

    const trainerItems = flattenItems(props.trainer)
      .map(formatTrainerItem)
      .filter((item) => Boolean(item.sound));

    const items = makeItemsWithDistractors(trainerItems);

    this.state = {
      items,
      currentIndex: 0,
      attempt: undefined,
      showChoiceItems: false,
      currentShuffledItems: shuffleItems(0, items)
    };

    this.handleContinueClick = this.handleContinueClick.bind(this);
    this.handleChoiceItemClick = this.handleChoiceItemClick.bind(this);
    this.onSoundPlayerEnded = this.onSoundPlayerEnded.bind(this);
    this.onSoundPlayerError = this.onSoundPlayerError.bind(this);
    this.handleToggleShortcuts = this.handleToggleShortcuts.bind(this);
    this.handleKeyboardTriggeredAttempt = this.handleKeyboardTriggeredAttempt.bind(this);
  }

  UNSAFE_componentWillMount() {
    const { items } = this.state;
    this.props.onStart({ scorableItemsCount: items.length });
  }

  handleToggleShortcuts() {
    const { toggleHints, track, locale, learnLanguageAlpha3 } = this.props;

    toggleHints();

    /* eslint-disable camelcase */
    track({
      event: 'gui:interacted',
      version: 1,
      payload: {
        gui_container: 'listening_trainer',
        gui_element: 'shortcut_hint_toggler',
        origin: 'lesson_player',
        interaction: 'clicked',
        locale,
        learn_language_alpha3: learnLanguageAlpha3
      }
    });
    /* eslint-enable camelcase */
  }

  handleKeyboardTriggeredAttempt() {
    const { track, locale, learnLanguageAlpha3 } = this.props;
    /* eslint-disable camelcase */
    track({
      event: 'gui:interacted',
      version: 1,
      payload: {
        gui_container: 'listening_trainer',
        gui_element: 'keyboard_shortcut',
        origin: 'lesson_player',
        interaction: 'clicked',
        locale,
        learn_language_alpha3: learnLanguageAlpha3
      }
    });
    /* eslint-enable camelcase */
  }

  handleChoiceItemClick(attempt, keyboardTriggered) {
    const { mediaUrlService, audio } = this.props;
    const { item } = this.state.items[this.state.currentIndex];
    const itemSoundUrl = mediaUrlService.soundURL(item.sound.id, 'mp3');

    this.setState({ attempt });

    if (keyboardTriggered) {
      this.handleKeyboardTriggeredAttempt();
    }

    const solved = attempt.mistakes === 0;
    const onEnded = () => {
      // Check if the item is still active
      const { item: currentItem } = this.state.items[this.state.currentIndex];
      if (item === currentItem && solved) {
        audio.playSoundWithState(itemSoundUrl);
      }
    };
    const options = {
      onEnded
    };

    playFeedbackSound(FEEDBACK_SOUND.forItemSolved(solved), options);

    this.props.attemptItem(item, {
      solved: attempt.mistakes === 0,
      text: attempt.text
    });
  }

  handleContinueClick(item, mistakes) {
    this.props.completeItem(item, mistakes);

    this.setState({ attempt: undefined, showChoiceItems: false });

    this.props.audio.stop();
    this.props.audio.cleanup();

    if (this.state.currentIndex < this.state.items.length - 1) {
      const nextIndex = this.state.currentIndex + 1;
      this.setState({
        currentIndex: nextIndex,
        currentShuffledItems: shuffleItems(nextIndex, this.state.items)
      });
    } else {
      setTimeout(() => this.props.onFinish(), 100);
    }
  }

  onSoundPlayerEnded() {
    this.setState({ showChoiceItems: true });
  }

  onSoundPlayerError() {
    const { items, currentIndex, attempt } = this.state;
    const { item: itemWithBrokenSound } = items[currentIndex];
    const { trainer, addMessage } = this.props;

    const trainerData = {
      interaction_mode: trainer.interaction,
      item_position_in_trainer: currentIndex + 1,
      number_of_items_in_trainer: items.length,
      translation_mode: ''
    };

    if (!navigator.onLine) {
      return addMessage(MESSAGE_KEYS.NETWORK);
    }

    this.props.skipItem(itemWithBrokenSound, attempt, trainerData);

    this.setState({ attempt: undefined, showChoiceItems: false });

    if (this.state.currentIndex < this.state.items.length - 1) {
      this.setState((prevState) => ({
        items: prevState.items.filter((itemWithDistractors) => itemWithDistractors.item !== itemWithBrokenSound)
      }));
    } else {
      this.props.onFinish();
    }
  }

  render() {
    const { attempt, showChoiceItems, currentShuffledItems } = this.state;
    const { item } = this.state.items[this.state.currentIndex];

    return (
      <ListeningTrainer
        key={item.id}
        mediaUrlService={this.props.mediaUrlService}
        choiceItemsList={currentShuffledItems}
        onSoundPlayerEnded={this.onSoundPlayerEnded}
        onSoundPlayerError={this.onSoundPlayerError}
        attempt={attempt}
        showChoiceItems={showChoiceItems}
        item={item}
        onContinue={this.handleContinueClick}
        onChoiceItemClick={this.handleChoiceItemClick}
        audio={this.props.audio}
        onToggleShortcuts={this.handleToggleShortcuts}
      />
    );
  }
}

const mapDispatchToProps = {
  track,
  toggleHints,
  attemptItem: sequenceActions.attemptItem,
  completeItem: sequenceActions.completeItem,
  skipItem: sequenceActions.skipItem,
  addMessage: messagesActions.addMessage
};

const mapStateToProps = ({ session }) => ({
  locale: session.locale,
  learnLanguageAlpha3: session.learnLanguageAlpha3
});

ListeningTrainerContainer.propTypes = {
  addMessage: PropTypes.func.isRequired,
  attemptItem: PropTypes.func.isRequired,
  completeItem: PropTypes.func.isRequired,
  skipItem: PropTypes.func.isRequired,
  trainer: PropTypes.object.isRequired,
  onStart: PropTypes.func.isRequired,
  onFinish: PropTypes.func.isRequired,
  mediaUrlService: PropTypes.any.isRequired,
  audio: audioPropTypes.isRequired,
  track: PropTypes.func.isRequired,
  locale: PropTypes.string.isRequired,
  learnLanguageAlpha3: PropTypes.string.isRequired,
  toggleHints: PropTypes.func.isRequired
};

export default compose(
  withServices(['mediaUrlService']),
  withAudio(),
  connect(mapStateToProps, mapDispatchToProps)
)(ListeningTrainerContainer);
