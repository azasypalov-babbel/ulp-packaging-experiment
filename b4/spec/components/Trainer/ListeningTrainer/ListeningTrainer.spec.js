import React from 'react';
import { shallow } from 'enzyme';
import { ListeningTrainer } from '../../../../src/components/Trainer/ListeningTrainer/ListeningTrainer';

jest.mock('../../../../src/components/shared/FeedbackSheet/FeedbackSheet', () => 'FeedbackSheet');

describe('<ListeningTrainer />', () => {
  const mockOnContinue = jest.fn();
  const mockOnChoiceItemClick = jest.fn();
  const mockAudio = {
    playSound: jest.fn(),
    playSoundWithState: jest.fn(),
    preload: jest.fn(),
    reset: jest.fn(),
    stop: jest.fn()
  };

  const defaultProps = {
    onUserTriggeredSound: jest.fn(),
    showChoiceItems: false,
    item: {
      id: '549f9ecf012e7f3b636ec9e4ac5a0d10',
      learnLanguageText: 'Uno',
      displayLanguageText: 'One',
      sound: {
        id: 'd3572193477785b1bc030cdf61ec777c'
      }
    },
    choiceItemsList: [
      {
        learnLanguageText: 'Dos',
        displayLanguageText: 'Two'
      },
      {
        learnLanguageText: 'Tres',
        displayLanguageText: 'Three'
      }
    ],
    mediaUrlService: {
      soundURL: jest.fn((val) => `http://localhost/${val}.mp3`)
    },
    onContinue: mockOnContinue,
    onChoiceItemClick: mockOnChoiceItemClick,
    translations: {
      attemptFeedback: {
        positiveFeedback: 'good',
        negativeFeedback: 'bad'
      },
      continueBtn: 'Continue',
      instruction: 'Listen to the phrase'
    },
    audio: mockAudio,
    onSoundPlayerEnded: () => {},
    onSoundPlayerError: () => {},
    onToggleShortcuts: jest.fn()
  };

  test('renders', () => {
    const component = shallow(<ListeningTrainer {...defaultProps} />);
    expect(component).toMatchSnapshot();
  });

  test('renders in attempted state', () => {
    const component = shallow(<ListeningTrainer
      {...defaultProps}
      attempt={{ text: 'One', mistakes: 0 }}
    />);

    expect(component).toMatchSnapshot();
  });

  describe('choiceItem click', () => {
    let component;
    let handleChoiceItemClick;

    beforeEach(() => {
      component = shallow(<ListeningTrainer {...defaultProps} />);
      handleChoiceItemClick = component.instance().handleChoiceItemClick;
      mockOnChoiceItemClick.mockClear();
    });

    describe('when correct', () => {
      test('calls onChoiceItemClick callback with expected attempt', () => {
        handleChoiceItemClick('One');

        const expectedAttempt = {
          text: 'One',
          mistakes: 0
        };

        expect(mockOnChoiceItemClick).toHaveBeenCalledWith(expectedAttempt, expect.anything());
      });
    });

    describe('when incorrect', () => {
      test('calls onChoiceItemClick callback with 2 mistakes', () => {
        handleChoiceItemClick('Two');

        const expectedAttempt = {
          text: 'Two',
          mistakes: 2
        };

        expect(mockOnChoiceItemClick).toHaveBeenCalledWith(expectedAttempt, expect.anything());
      });
    });
  });

  describe('continue button click', () => {
    let component;
    let handleContinueClick;

    const mockOnContinue = jest.fn();

    describe('when attempt exist', () => {
      const attempt = {
        mistakes: 1,
        text: 'Two'
      };

      beforeEach(() => {
        component = shallow(<ListeningTrainer
          {...defaultProps}
          attempt={attempt}
          onContinue={mockOnContinue}
        />);
        handleContinueClick = component.instance().handleContinueClick;
      });

      test('calls onContinue callback', () => {
        handleContinueClick();

        expect(mockOnContinue).toHaveBeenCalledTimes(1);
        expect(mockOnContinue).toHaveBeenCalledWith(defaultProps.item, attempt.mistakes);
      });
    });

    describe('when attempt is reset to undefined', () => {
      const attempt = undefined;

      beforeEach(() => {
        component = shallow(<ListeningTrainer
          {...defaultProps}
          attempt={attempt}
          onContinue={mockOnContinue}
        />);
        handleContinueClick = component.instance().handleContinueClick;
      });

      test('does not call onContinue callback', () => {
        handleContinueClick();

        expect(mockOnContinue).not.toHaveBeenCalled();
      });
    });
  });

  describe('Toolbar', () => {
    let component;

    test('renders', () => {
      component = shallow(<ListeningTrainer {...defaultProps} />);

      expect(component.find('Toolbar')).toHaveLength(1);
    });

    test('does not render when attempt', () => {
      component = shallow(<ListeningTrainer
        attempt={{ text: 'One', mistakes: 0 }}
        {...defaultProps}
      />);

      expect(component.find('Toolbar')).toHaveLength(0);
    });
  });

  describe('with attempt', () => {
    let component;

    test('renders FeedbackSheet', () => {
      component = shallow(<ListeningTrainer {...defaultProps} />);
      expect(component.find('FeedbackSheet')).toHaveLength(0);
      component.setProps({ attempt: { text: 'Two', mistakes: 1 } });
      expect(component.find('FeedbackSheet')).toHaveLength(1);
    });
  });
});
