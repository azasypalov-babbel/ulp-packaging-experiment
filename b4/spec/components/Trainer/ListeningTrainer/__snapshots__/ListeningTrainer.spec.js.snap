// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListeningTrainer /> renders 1`] = `
<div>
  <TrainerLayout
    bottom={
      <ChoiceList
        correctText="One"
        items={
          Array [
            "Two",
            "Three",
          ]
        }
        onClick={[Function]}
        showChoiceItems={false}
      />
    }
    top={
      <React.Fragment>
        <Title
          as="h1"
          color="spaceGray"
          data-selector="listening-title"
          fontFamily="fontMilliardSemi"
          fontSize="big"
          textAlign="center"
        >
          Listen to the phrase
        </Title>
        <SoundPlayer
          audio={
            Object {
              "playSound": [MockFunction],
              "playSoundWithState": [MockFunction],
              "preload": [MockFunction],
              "reset": [MockFunction],
              "stop": [MockFunction],
            }
          }
          onEnded={[Function]}
          onError={[Function]}
          playOnMount={true}
          url="http://localhost/d3572193477785b1bc030cdf61ec777c.mp3"
        />
      </React.Fragment>
    }
  />
  <Toolbar>
    <WithKeypress(IconButton)
      data-selector="reveal-shortcuts-button"
      iconName="KeyboardIcon"
      onClick={[MockFunction]}
    />
  </Toolbar>
</div>
`;

exports[`<ListeningTrainer /> renders in attempted state 1`] = `
<div>
  <TrainerLayout
    bottom={
      <ChoiceList
        attemptedText="One"
        correctText="One"
        items={
          Array [
            "Two",
            "Three",
          ]
        }
        onClick={[Function]}
        showChoiceItems={false}
      />
    }
    top={
      <React.Fragment>
        <Title
          as="h1"
          color="spaceGray"
          data-selector="listening-title"
          fontFamily="fontMilliardSemi"
          fontSize="big"
          textAlign="center"
        >
          Listen to the phrase
        </Title>
        <SoundPlayer
          audio={
            Object {
              "playSound": [MockFunction],
              "playSoundWithState": [MockFunction],
              "preload": [MockFunction],
              "reset": [MockFunction],
              "stop": [MockFunction],
            }
          }
          onEnded={[Function]}
          onError={[Function]}
          playOnMount={true}
          url="http://localhost/d3572193477785b1bc030cdf61ec777c.mp3"
        />
      </React.Fragment>
    }
  />
  <FeedbackSheet
    data-gap-feedback={true}
    data-mistaken={false}
    data-solved={true}
    feedbackDetail={
      <div
        data-selector="learn-language-text"
      >
        <HighlightedText
          appearance="POSITIVE"
        >
          Uno
        </HighlightedText>
      </div>
    }
    feedbackMessage="good"
    key="feedback-sheet"
    onContinueClick={[Function]}
  />
</div>
`;
